[cdac] Implement a JSON contract reader

lambdageek opened the pull request:
Contributes to #99298 

Implement a parser for the "compact" JSON contract descriptor format specified in [data_contracts.md](https://github.com/dotnet/runtime/blob/main/docs/design/datacontracts/data_descriptor.md)

The data model is a WIP - it's likely we will want something a bit more abstract (and less mutable).

This is not wired up to consume a real contract descriptor from target process memory at the moment.  It's only exercised by unit tests for now.

TODO:

- [x] hook up the unit tests to some testing infrastructure - either `src/libraries` or `src/tests` or to something else

lambdageek committed:
WIP: compact descriptor format json parser

lambdageek committed:
checkpoint: camelCase dict keys; compact fields

lambdageek committed:
more parsing

am11 commented on src/coreclr/tools/cdac-build-tool/Program.cs:
```diff
@@ -0,0 +1,42 @@
+using System.Text.Encodings.Web;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
```
nit: delete unused usings on C# sources to reduce the noise.

lambdageek replied:
@am11 thanks.  this change is actually part of #100650 (PTAL if you're interested) - and I already changed that one to remove the unused usings.  I'll rebase this current PR so it doesn't include changes from that one.  I thought I might end up sharing code, but it didn't seem useful yet.

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/ContractDescriptorParser.cs:
```diff
@@ -0,0 +1,318 @@
+namespace Microsoft.Diagnostics.DataContractReader;
+public partial class ContractDescriptorParser
+{
+    public const string TypeDescriptorSizeSigil = "!";
```
If this is part of the specification, can we get a link?

lambdageek committed:
Follow C# coding guidelines

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

lambdageek committed:
additional comments

am11 commented on src/native/managed/cdacreader/src/ContractDescriptorParser.cs:
```diff
@@ -0,0 +1,326 @@
+///   A parser for the JSON representation of a contract descriptor.
+/// </summary>
+/// <remarks>
+/// See docs/design/datacontracts/data_descriptor.md for the format
```
nit:
```suggestion
/// <see href="https://github.com/dotnet/runtime/blob/main/docs/design/datacontracts/data_descriptor.md">See design doc</see> for the format.
```

lambdageek commented:
By the way, the unit tests for cdacreader aren't hooked up to any CI infrastructure at the moment.  I'm not sure if we should include them into `src/tests` or `src/libraries` or directly into some yaml pipeline.  Open to suggestions.

lambdageek committed:
better doc comments

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

jkoritzinsky commented:
For hooking up the unit tests, I think hooking it up in the same way as the ILCompiler and ILLink unit tests will be the easiest way.

That way it won't take on additional dependencies or complexity introduced by the other test harnesses.

lambdageek commented on src/native/managed/cdacreader/tests/Directory.Build.props:
```diff
@@ -0,0 +1,3 @@
+<Project>
+  <Import Project="..\..\..\..\libraries\Directory.Build.props" />
```
`dotnet.sh build ... /t:Test` doesnt' seem to work without _something_ from libraries.   It's probably `eng/testing/tests.props`, but I'm not sure.  I'm cargo-culting from how the FunctionalTests are set up.

lambdageek commented:
> For hooking up the unit tests, I think hooking it up in the same way as the ILCompiler and ILLink unit tests will be the easiest way.

@jkoritzinsky just so I'm clear: I would need to add a new lane in the yaml file, right?

lambdageek commented on src/native/managed/cdacreader/src/ContractDescriptorParser.cs:
```diff
@@ -0,0 +1,326 @@
+    internal sealed class FieldDescriptorConverter : JsonConverter<FieldDescriptor>
+    {
+        // Compact Field descriptors are either one or two element arrays
+        // 1. [number] - no type, offset is given as the number
```
Oops. You're right. There are just two cases: `number` or `[number, string]`. The converter also handled `[number]` which is not valid.  I'm removing it.

jkoritzinsky commented:
You could add a subset and add that subset to the CLR Tools Tests job's build args (and update the path triggers). Then it will run as part of that job. No need for an additional job.

lambdageek committed:
suggestions from reviews; remove FieldDescriptor wrong conversion

we incorrectly allowed `[number]` as a field descriptor
conversion. that's not allowed.  removed it.

lambdageek committed:
Make test project like the nativeoat+linker tests

Dont' use libraries test infrastructure.  Just normal arcade xunit support.

lambdageek committed:
add tools.cdacreadertests subset; add to CLR_Tools_Tests test leg

lambdageek commented:
Added a `tools.cdacreadertests` subset - added it to the `CLR_Tools_Tests` test job

lambdageek committed:
one more code review chang

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -369,6 +369,10 @@
       Test="true" Category="clr" Condition="'$(DotNetBuildSourceOnly)' != 'true' and '$(NativeAotSupported)' == 'true'"/>
   </ItemGroup>
 
+  <ItemGroup Condition="$(_subset.Contains('+tools.cdacreadertests+'))">
```
You also need to declare the subset above with a `SubsetName` item (that will fix the validation error and also provide some basic docs for the subset)

lambdageek committed:
add tools.cdacreadertests for subset validation

elinor-fung reviewed (commented):
Are there docs anywhere about tools tests (adding, running, etc) that we should update?

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
reorder cases

lambdageek commented:
> Are there docs anywhere about tools tests (adding, running, etc) that we should update?

Added a new PR with docs: https://github.com/dotnet/runtime/pull/101186

lambdageek committed:
no duplicate fields/sizes in types

lambdageek commented:
> ```
> .dotnet/sdk/9.0.100-preview.3.24204.13/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: (NETCORE_ENGINEERING_TELEMETRY=Build) Assets file '/__w/1/s/artifacts/obj/cdacreader/project.assets.json' doesn't have a target for 'net9.0/linux-x64'. Ensure that restore has run and that you have included 'net9.0' in the TargetFrameworks for your project. You may also need to include 'linux-x64' in your project's RuntimeIdentifiers.
> ```

there's a related  failure on the "CLR_Tools_Tests" lane that is similar to what we previously saw with the `src/native/managed` infrastructure - it's something about the right properties not flowing down to the Restore target, I think.

elinor-fung reviewed (approved):
Looks good to me - assuming whatever is going on with the CLR_Tools_Tests leg is straightforward to address.

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-parse-json

lambdageek committed:
wip

lambdageek committed:
Make all cdacreader.csproj ProjectReferences use the same AdditionalProperties

Since we set Configuration and RuntimeIdentifier, if we don't pass the
same AdditionalProperties in all ProjectReferences, we bad
project.assets.json files

lambdageek commented:
Ah. The issue was that we now have two `<ProjectReference Include="cdacreader.csproj">` items.  One in compile-native.proj and the other in the unit test.  But they had distinct `AdditionalProperties` metadata, and the one in `compile-native.proj` set `Configuration` and `RuntimeIdentifier`, whereas the test didn't.  as a result the restore created a project.assets.json that was only correct for one reference and not the other.  the fix is to set the same `AdditionalProperties` on both references.

We set `RuntimeIdentifier=$(OutputRID)` so that in the case of cross-arch build NativeAOT will target the target architecture.  but I wonder if there's a better way to do that than by messing with ProjectReference - which is apparently kind of fragile.

lambdageek committed:
Don't share the native compilation AdditionalProperties

lambdageek commented on src/native/managed/subproject.props:
```diff
@@ -0,0 +1,19 @@
+        <SubprojectProps Include="LibrariesConfiguration" Value="$(LibrariesConfiguration)" />
+        <SubprojectProps Include="RuntimeIdentifier" Value="$(OutputRID)" />
+
+        <SubprojectProps Include="NativeLib" Value="$(NativeLibKind)" />
```
Good catch. there are a couple of these `AdditionalProperties` that are only relevant when we're doing a proper native compilation.  I moved those back to compile-native.proj
